File: src/main/java/org/gd/leetcode/p0394/solution.py
class Solution:

    def __init__(self):
        super().__init__()
        self.sb: str = ""

    def decodeString(self, s: str) -> str:
        if not s:
            return s

        self.sb: str = ""

        def find_last(start: int, end: int) -> int:
            i: int = start
            open_count: int = 1
            while i < end:
                if s[i] == '[':
                    open_count += 1
                elif s[i] == ']':
                    open_count -= 1
                if open_count == 0:
                    return i
                i += 1

            return -1

        def decode_string(start: int, end: int):
            i: int = start
            d: int = 0
            while i < end:
                c = s[i]
                if c == ']':
                    continue
                if c.isdigit():
                    d = d * 10 + int(c)
                elif c == '[':
                    last = find_last(i + 1, end)
                    while d > 0:
                        decode_string(i + 1, last)
                        d -= 1
                    i = last
                else:
                    self.sb += c

                i += 1

        decode_string(0, len(s))

        return self.sb

File: src/main/java/org/gd/leetcode/p0394/Solution.java
package org.gd.leetcode.p0394;

import org.gd.leetcode.common.LeetCode;

/**
 * https://leetcode.com/problems/decode-string/
 */
@LeetCode(
        name = "Decode String",
        state = LeetCode.State.DONE,
        difficulty = LeetCode.Level.MEDIUM,
        tags = {
                LeetCode.Tags.STACK,
                LeetCode.Tags.DEPTH_FIRST_SEARCH
        }
)
class Solution {

    private StringBuilder sb;

    public String decodeString(String s) {

        if (s == null) {
            return null;
        }

        if (s.isEmpty()) {
            return "";
        }

        sb = new StringBuilder();
        _decodeString(s, 0, s.length());
        return sb.toString();
    }

    private static int findLast(String s, int startIndex, int endIndex) {
        int open = 1;
        for (int i = startIndex; i < endIndex; i++) {
            char c = s.charAt(i);
            if (c == '[') {
                open++;
            } else if (c == ']') {
                open--;
            }
            if (open == 0) {
                return i;
            }
        }
        throw new IllegalArgumentException("input string is not valid ( " + s + " )");
    }

    private void _decodeString(String s, int startIndex, int endIndex) {

        if (endIndex - startIndex <= 0) {
            return;
        }

        if (endIndex - startIndex == 1) {
            sb.append(s.charAt(startIndex));
            return;
        }

        int digit = 0;
        for (int i = startIndex; i < endIndex; i++) {
            char c = s.charAt(i);
            if (c == ']') {
                continue;
            }
            if (Character.isDigit(c)) {
                digit = digit * 10 + Character.digit(c, 10);
            } else if (c == '[') {
                int last = findLast(s, i + 1, endIndex);
                while (digit > 0) {
                    _decodeString(s, i + 1, last);
                    digit--;
                }
                i = last;
            } else {
                sb.append(c);
            }
        }
    }
}
